From 64245a20b78824634008b240c3f97c29a353ebe3 Mon Sep 17 00:00:00 2001
From: Karmjit Mahil <karmjit.mahil@igalia.com>
Date: Mon, 19 Oct 2026 11:41:20 +0100
Subject: [PATCH] tu: Add back float16 3D path for multisampled copies

A2D can't handle multisampled copies so we need the 3D path here,
and the float16 specific path to preserve NaNs.

This reverts commit 6237388454045996f58719ad33cd943e55e26a64

Signed-off-by: Karmjit Mahil <karmjit.mahil@igalia.com>
Part-of: <https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/45366>
---
 src/freedreno/vulkan/tu_clear_blit.cc | 38 ++++++++++++++++++++-------
 src/freedreno/vulkan/tu_device.h      |  1 +
 2 files changed, 30 insertions(+), 9 deletions(-)

diff --git a/src/freedreno/vulkan/tu_clear_blit.cc b/src/freedreno/vulkan/tu_clear_blit.cc
index 714b103b843..4b611ba0eeb 100644
--- a/src/freedreno/vulkan/tu_clear_blit.cc
+++ b/src/freedreno/vulkan/tu_clear_blit.cc
@@ -997,7 +997,7 @@ build_blit_fs_shader(bool zscale, bool swap_coords)
  * variant for them.
  */
 static nir_shader *
-build_ms_copy_fs_shader(void)
+build_ms_copy_fs_shader(bool half_float)
 {
    nir_builder _b =
       nir_builder_init_simple_shader(MESA_SHADER_FRAGMENT, NULL,
@@ -1006,7 +1006,8 @@ build_ms_copy_fs_shader(void)
 
    nir_variable *out_color =
       nir_create_variable_with_location(b->shader, nir_var_shader_out,
-                                        FRAG_RESULT_DATA0, glsl_vec4_type());
+                                        FRAG_RESULT_DATA0,
+                                        half_float ? glsl_f16vec_type(4) : glsl_vec4_type());
    nir_variable *in_coords =
       nir_create_variable_with_location(b->shader, nir_var_shader_in,
                                         VARYING_SLOT_VAR0, glsl_vec_type(2));
@@ -1024,7 +1025,7 @@ build_ms_copy_fs_shader(void)
                              /* Note: since we're just copying data, we rely on
                               * the HW ignoring the dest_type.
                                */
-                             .dest_type = nir_type_int32);
+                             .dest_type = half_float ? nir_type_float16 : nir_type_int32);
 
    nir_store_var(b, out_color, tex, 0xf);
    return b->shader;
@@ -1139,7 +1140,8 @@ tu_init_clear_blit_shaders(struct tu_device *dev)
    compile_shader(dev, build_blit_fs_shader(true, false), 0, &offset, GLOBAL_SH_FS_BLIT_ZSCALE);
    compile_shader(dev, build_blit_fs_shader(false, true), 0, &offset, GLOBAL_SH_FS_BLIT_SWAP_COORDS);
    compile_shader(dev, build_blit_fs_shader(true, true), 0, &offset, GLOBAL_SH_FS_BLIT_ZSCALE_SWAP_COORDS);
-   compile_shader(dev, build_ms_copy_fs_shader(), 0, &offset, GLOBAL_SH_FS_COPY_MS);
+   compile_shader(dev, build_ms_copy_fs_shader(false), 0, &offset, GLOBAL_SH_FS_COPY_MS);
+   compile_shader(dev, build_ms_copy_fs_shader(true), 0, &offset, GLOBAL_SH_FS_COPY_MS_HALF);
    for (unsigned i = 0; i < 3; i++) {
       compile_shader(dev, build_ms_resolve_fs_shader(2 << i), 0, &offset,
                      static_cast<enum global_shader>(GLOBAL_SH_FS_RESOLVE_MS2 + i));
@@ -1164,6 +1166,7 @@ enum r3d_type {
    R3D_CLEAR,
    R3D_BLIT,
    R3D_MULTI_BLIT,
+   R3D_COPY_HALF,
 };
 
 template <chip CHIP>
@@ -1183,6 +1186,9 @@ r3d_common(struct tu_cmd_buffer *cmd, struct tu_cs *cs, enum r3d_type type,
 
    if (z_scale) {
       fs_id = swap_coords ? GLOBAL_SH_FS_BLIT_ZSCALE_SWAP_COORDS : GLOBAL_SH_FS_BLIT_ZSCALE;
+   } else if (type == R3D_COPY_HALF) {
+      /* Avoid canonicalizing NaNs due to implicit conversions in the shader. */
+      fs_id = GLOBAL_SH_FS_COPY_MS_HALF;
    } else if (swap_coords) {
       fs_id = GLOBAL_SH_FS_BLIT_SWAP_COORDS;
    } else if (src_samples != VK_SAMPLE_COUNT_1_BIT) {
@@ -1302,7 +1308,8 @@ r3d_common(struct tu_cmd_buffer *cmd, struct tu_cs *cs, enum r3d_type type,
          unsigned regid = 0;
          if (rts_mask & (1u << i))
             regid = ir3_find_output_regid(fs, FRAG_RESULT_DATA0 + rt++);
-         tu_cs_emit(cs, A6XX_SP_PS_OUTPUT_REG_REGID(regid));
+         tu_cs_emit(cs, A6XX_SP_PS_OUTPUT_REG_REGID(regid) |
+                        COND(regid & HALF_REG_ID, A6XX_SP_PS_OUTPUT_REG_HALF_PRECISION));
       }
    }
 
@@ -1927,8 +1934,18 @@ r3d_setup(struct tu_cmd_buffer *cmd,
       }
    }
 
-   const enum r3d_type type = (clear) ? R3D_CLEAR :
-      ((blit_param & R3D_USE_MULTI_BLIT) ? R3D_MULTI_BLIT : R3D_BLIT);
+   enum r3d_type type;
+   if (clear) {
+      type = R3D_CLEAR;
+   } else if ((blit_param & R3D_COPY) && util_format_is_float16(src_format) &&
+              dst_samples != VK_SAMPLE_COUNT_1_BIT) {
+      /* Avoid canonicalizing NaNs in copies by using the special half-float
+       * path that uses half regs.
+       */
+      type = R3D_COPY_HALF;
+   } else {
+      type = (blit_param & R3D_USE_MULTI_BLIT) ? R3D_MULTI_BLIT : R3D_BLIT;
+   }
    r3d_common<CHIP>(cmd, cs, type, 1, blit_param & R3D_Z_SCALE, src_samples,
                     dst_samples, blit_param & R3D_SWAP_COORDS);
 
@@ -3388,8 +3405,9 @@ tu_copy_image_to_image(struct tu_cmd_buffer *cmd,
 {
    const struct blit_ops *ops = &r2d_ops<CHIP>;
    struct tu_cs *cs = &cmd->cs;
+   const bool is_msaa = dst_image->layout[0].nr_samples > 1;
 
-   if (dst_image->layout[0].nr_samples > 1)
+   if (is_msaa)
       ops = &r3d_ops<CHIP>;
 
    VkOffset3D src_offset = info->srcOffset;
@@ -3424,7 +3442,9 @@ tu_copy_image_to_image(struct tu_cmd_buffer *cmd,
    /* note: could use "R8_UNORM" when no UBWC */
    unsigned blit_param = 0;
    if (dst_format == PIPE_FORMAT_Y8_UNORM ||
-       src_format == PIPE_FORMAT_Y8_UNORM) {
+       src_format == PIPE_FORMAT_Y8_UNORM ||
+       (is_msaa && (util_format_is_float16(src_format) ||
+                    util_format_is_float16(dst_format)))) {
       ops = &r3d_ops<CHIP>;
       blit_param = R3D_COPY;
    }
diff --git a/src/freedreno/vulkan/tu_device.h b/src/freedreno/vulkan/tu_device.h
index 5ef943a684d..cf48f0d1831 100644
--- a/src/freedreno/vulkan/tu_device.h
+++ b/src/freedreno/vulkan/tu_device.h
@@ -52,6 +52,7 @@ enum global_shader {
    GLOBAL_SH_FS_BLIT_SWAP_COORDS,
    GLOBAL_SH_FS_BLIT_ZSCALE_SWAP_COORDS,
    GLOBAL_SH_FS_COPY_MS,
+   GLOBAL_SH_FS_COPY_MS_HALF,
    GLOBAL_SH_FS_RESOLVE_MS2,
    GLOBAL_SH_FS_RESOLVE_MS4,
    GLOBAL_SH_FS_RESOLVE_MS8,
-- 
2.43.0

